server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java

```diff
@@ -304,29 +313,21 @@ private static HitContext prepareNonNestedHitContext(
                 hit = new SearchHit(docId, leafStoredFieldLoader.id(), emptyMap(), emptyMap());
             }
 
-            HitContext hitContext = new HitContext(hit, subReaderContext, subDocId);
-            BytesReference source;
+            Source source;
             if (sourceRequired(context)) {
                 try {
                     profiler.startLoadingSource();
-                    source = sourceLoader.source(leafStoredFieldLoader, subDocId);
+                    source = Source.fromBytes(sourceLoader.source(leafStoredFieldLoader, subDocId));
+                    SourceLookup scriptSourceLookup = context.getSearchExecutionContext().lookup().source();
+                    scriptSourceLookup.setSegmentAndDocument(subReaderContext, subDocId);
+                    scriptSourceLookup.setSourceProvider(new SourceLookup.BytesSourceProvider(source.internalSourceRef()));
                 } finally {
                     profiler.stopLoadingSource();
                 }
             } else {
-                source = null;
+                source = Source.EMPTY;
```

Comment:
Hmm, sometimes source is needed in various places in the fetch phase (like script fields that load from source), but we don't load it up front. I think that `sourceRequired` is really just a best-effort to improve performance, since it's preferable to load all stored fields up-front.

Concretely, let's say you had search request with `highlight` but `_source: false`. Would highlighting still be able to load source correctly?

Reply:
That's a good point.  Highlighting is actually called out specifically as something that inner hits has to take into account because it can't rely on the SourceLookup returning the correct document.  Scripts would continue to work because we only set the source provider on the SearchLookup if we've loaded source up-front, and from a quick inspection I think script fields that reference source in inner hits don't actually work at the moment because we only set it for non-nested hits.

Maybe a better way of doing this would be to have a method on FetchSubPhaseProcessor that indicates whether or not the sub phase requires source? I'll give it a go and see if that works.

Reply:
I've replaced `Source.EMPTY` with a lazy-loading source implementation instead, which means we can remove the extra checks in `requiresSource`.  I think having a 'what do we need to load' method on FetchSubPhaseProcessor would be helpful though, so that's something to look at next.

I realise I should clarify my comment about inner hits loading source.  I agree that allowing inner hits to load source when you don't want it on the top level is useful, what I find odd is that if you disable it at the top level you don't have to specifically enable it lower down to get the source there.  It feels backwards to me that you have to set `enabled:false` at both levels to make sure that you're getting the performance improvement.

Reply:
> what I find odd is that if you disable it at the top level you don't have to specifically enable it lower down to get the source there

Got it, I see how this is extra work for users. But I also understand why it's like this -- the configuration within `inner_hits` is completely separate from the top-level configuration.

Reply:
Well sort of.  For most things it's actually the same - the inner hits search context wraps the top level search context and delegates to it for most things, but there are a few bits that it overloads (fetch source and fetch fields being two of them) - it's all very confusing!